# Análisis datos Airbnb Seattle - Minería de datos y el paradigma Big Data
__Autores:__
- Alberto Pastor Moreno
- Iván Fernández Mena
- Francisco Javier Melchor González

## Fase I. Business Understanding. Definición de necesidades del cliente. Comprensión del negocio.

### Objetivos del negocio

El principal objetivo del proyecto se enfoca en la resolución de la siguiente cuestión: dado un perfil host de Airbnb, un apartamento perteneciente a éste en Seattle y el mapa de espacios culturales tematizados en esta ciudad, ¿qué precio por noche aconseja nuestro sistema?

### Valoración de la situación

Este análisis utiliza datos sobre [la actividad de Airbnb en Seattle](https://www.kaggle.com/airbnb/seattle) y [espacios culturales en Seattle](https://data.seattle.gov/Community/Seattle-Cultural-Space-Inventory/hrat-v8cm)

### Objetivos de Data Mining

### Plan de proyecto

## Fase II. Data Understanding. Estudio y comprensión de los datos

### Recolectar los datos iniciales

La solución a la cuestión planteada, concentrada en el análisis predictivo, hace uso de técnicas de cruce de varios conjuntos de datos. Los conjuntos de datos abiertos utilizados a lo largo del proceso que hemos creído relevantes y nutritivos para el objetivo del proyecto se pueden encontrar a continuación: 

Los datos a utilizar en este proyecto y sus fuentes son los siguientes:
- Listado de _hosts_ de Airbnb de Seattle, calendario de reservas de estos _hosts_ y las consecuentes _reviews_ de estos hosts, extraídos de [Kaggle](https://www.kaggle.com/airbnb/seattle)
- Listado de espacios culturales de Seattle, extraídos ha sido extraído del [portal de datos abiertos de esta ciudad](https://data.seattle.gov/Community/Seattle-Cultural-Space-Inventory/hrat-v8cm)

En los siguientes apartados analizaremos meticulosamente los datos del proyecto con el objetivo de realizar una comprensión exhausta de los datos elegidos para resolver la cuestión de este proyecto.

Cargamos todos los datos contenidos en formato `.csv` descargados en local previamente.

In [1]:
airbnb_listings_filename <- "datasets//listings.csv"
airbnb_calendar_filename <- "datasets//calendar.csv"
airbnb_reviews_filename <- "datasets//reviews.csv"
cultural_spaces_filename <- "datasets//cultural_spaces.csv"

In [2]:
airbnb_listings <- read.csv(file=airbnb_listings_filename, header=TRUE, sep=",")
airbnb_calendar <- read.csv(file=airbnb_calendar_filename, header=TRUE, sep=",")
airbnb_reviews <- read.csv(file=airbnb_reviews_filename, header=TRUE, sep=",")
cultural_spaces <- read.csv(file=cultural_spaces_filename, header=TRUE, sep=",")

### Descripción de los datos 

Para profundizar en la composición de los conjuntos de datos a tratar en este proyecto, analizaremos los atributos de cada uno de los conjuntos de datos cargados a continuación

#### Casas de Airbnb en Seattle

Primero, analizaremos el conjunto de datos `airbnb_listings`, el cual expone por cada fila la información completa que procesa de un _listing_ perteneciente a un _host_ de Seattle Airbnb, siendo esta la siguiente:
-  `	Id: ` Identificador del Airbnb
-  `	Listing_url: ` Página del Airbnb, dónde podemos encontrar fotos, fechas, comentarios, contactos con el propietario/a etc.
-  `	Scrape_id: ` Identificador de reserva
-  `	Last_scraped: ` Última reserva
-  `	Name: ` Nombre del Airbnb
-  `	Summary: ` Breve resumen del Airbnb en concreto que contiene características del mismo, actividades que pueden realizarse dependiendo de la situación del mismo, número de personas recomendado, localización, fecha en la que fue construido etc.
-  `	Space: ` Descripción del espacio del Airbnb, en algunas ocasiones viene el mismo contenido que en el atributo “Summary”, pero suelen incluir más detalles, así como características del paisaje que rodea el Airbnb, si hay o no restaurante o tiendas cercanos, si es tranquila o muy transitado el lugar donde se sitúa el Airbnb.
-  `	Description: ` Descripción detallada del Airbnb, que suele ser muy parecida o igual que el atributo “Summary” del mismo.
-  `	Experiences offered: ` Experiencias que ofrece el Airbnb, en todos aparece “nada”
-  `	Neighborhood_overview: ` Descripción o vision general del barrio donde se encuentra el Airbnb en concreto
-  `	Notes: ` Notas sobre el Airbnb en concreto, normalmente suelen incluir detalles que llamen la atención a los clientes para convencerlos de que lo alquilen o aparecen mensajes que indican que se pongan en contacto con el comprador si el cliente tiene alguna duda.
-  `	Transit: ` Este atributo indica, las diversas opciones de transporte público disponibles, que sean accesibles y que estén cerca del Airbnb.
-  `	Thumbnail_url: ` URL que nos lleva a una foto en miniatura del Airbnb
-  `	Medium_url: ` URL que nos lleva a la misma foto que el atributo anterior, pero un poco más grande
-  `	Picture_url: ` URL que nos lleva a la misma foto que los dos atributos anteriores, pero con un tamaño estándar
-  `	Xl_picture_url: ` URL con la misma foto que los 3 atributos anteriores, pero en tamaño grande.
-  `	Host_url: ` URL que conduce a una página donde aparece una descripción del usuario que alquila el Airbn, así como los anuncios que tiene, idiomas en los que habla, lugar donde vive, las guías de dicho usuario (si las tiene), comentarios, valoraciones, si ha confirmado número de teléfono, correo electrónico, correo electrónico de empresa y si tiene documentación oficial de identificación, nivel de usuario, cantidad de valoraciones que le han hecho los diferentes clientes, foto etc.
-  `	Host_name: ` Nombre del usuario que alquila el Airbnb
-  `	Host_since: ` Fecha desde la cual lleva registrada el usuario que alquila el Airbnb
-  `	Host_location: ` Lugar dónde vive el usuario
-  `	Host_about: ` Breve descripción sobre el usuario
-  `	Host_response_time: ` Disponibilidad del usuario para responder llamadas, correos o mensajes.
-  `	Host_response_rate: ` Porcentaje de respuesta del usuario
-  `	Host_acceptance_reate: ` Porcentaje de aceptación del usuario
-  `	Host_is_superhost: ` Indica ‘t’ si el usuario es superusuario o premium y ‘f’ si no lo es.
-  `	host_thumbnail_url: ` URL de una foto en miniatura que representa al usuario propietario
-  `	host_picture_url: ` URL de una foto en tamaño normal que representa al usuario propietario.
-  `	host_neighbourhood: ` Barrio del usuario propietario
-  `	host_listings_count: ` *
-  `	host_total_listings_count: ` *
-  `	host_verifications: ` Lista de verificaciones del usuario propietario, así como móvil, email, valoraciones, linkedin, Facebook 
-  `	host_has_profile_pic
-  `	host_identity_verified: ` Indica si la identidad del usuario está verificada, entendiendo ‘t’ como verificada y ‘f’ como no verificada.
-  `	Street: ` Dirección del Airbnb 
-  `	Neighbourhood: ` Barrio donde está situado el Airbnb
-  `	neighbourhood_cleansed: ` Zona concreta del barrio donde está situado el Airbnb
-  `	neighbourhood_group_cleansed: ` Recoge la misma información que el atributo “Neighbourhood”
-  `	city: ` Ciudad donde se encuentra el Airbnb
-  `	state: ` Estado donde se encuentra el Airbnb
-  `	zipcode: ` Código postal del Airbnb
-  `	market: ` *
-  `	smart_location: ` Nombre de la ciudad y estado
-  `	country_code: ` Abreviatura del país del Airbnb
-  `	country: ` País del airbnb
-  `	latitude: ` Latitud en la que se encuentra el Airbnb
-  `	longitude: ` Longitud en la que se encuentra el Airbnb
-  `	is_location_exact: ` Indica si es la localización exacta del Airbnb
-  `	property_type: ` Tipo de Airbnb, puede ser, piso, apartamento, chalé etc.
-  `	room_type: ` Tipo de alquiler, esto quiere decir si el alquiler es por habitaciones o de todo el espacio que ofrece el tipo de Airbnb que sea.
-  `	Accommodates: ` Habitaciones
-  `	Bathrooms: ` Cuartos de baño del Airbnb
-  `	Bedrooms: ` Habitaciones del Airbnb
-  `	Beds: ` Camas del Airbnb
-  `	bed_type: ` Tipos de camas que ofrece el Airbnb
-  `	amenities: ` Electrodomésticos que ofrece el Airbnb 
-  `	square_feet: ` metros cuadrados, suele aparecer vacía en la mayoría
-  `	Price: ` Precio por noche del Airbnb
-  `	weekly_price: ` Precio por semana del Airbnb
-  `	monthly_price: ` Precio por mes del Airbnb
-  `	security_deposit: ` Fianza para guardar la reserva
-  `	cleaning_fee: ` Precio de servicio de limpieza
-  `	guests_included: ` Invitados que permite traer al Airbnb
-  `	extra_people: ` Precio por persona extra al Airbnb
-  `	minimum_nights: ` Mínimo de noches que se deben reservar para una estancia.
-  `	maximum_nights: ` Máximo de noches que se pueden reservar.
-  `	calendar_updated: ` Última modificación del calendario
-  `	has_availability: ` Indica si está disponible para alquilar el Airbnb 
-  `	availability_30: ` Disponibilidad del Airbnb en un mes 
-  `	availability_60: ` Disponibilidad del Airbnb en dos meses 
-  `	availability_90: ` Disponibilidad del Airbnb en 3 meses 
-  `	availability_365: ` Disponibilidad del Airbnb en 1 año 
-  `	calendar_last_scraped: ` Fecha de la última reserva del Airbnb
-  `	number_of_reviews: ` Número de valoraciones que posee el Airbnb
-  `	first_review: ` Fecha de la primera valoración del Airbnb
-  `	last_review: ` Fecha de la última valoración del Airbnb
-  `	review_scores_rating: ` Media de puntuaciones en general del Airbnb
-  `	review_scores_accuracy: ` Media de puntuaciones en las reseñas con respecto a la medida en la que la página del anuncio coincide con el alojamiento
-  `	review_scores_cleanliness: ` Media de puntuaciones en las reseñas con respecto a la limpieza del Airbnb.
-  `	review_scores_checkin: ` Media de puntuación con respecto al trato al trato recibido con respecto a la llegada.
-  `	review_scores_communication: ` Media de puntuaciones en las reseñas con respecto a la medida en la que se ha tenido comunicación antes y después de la estancia con el usuario propietario.
-  `	review_scores_location: ` Media de puntuaciones en las reseñas con respecto a los huéspedes del barrio
-  `	review_scores_value: ` Media de puntuaciones en las reseñas con respecto a calidad-precio del Airbnb
-  `	requires_license: ` Indica si el Airbnb necesita licencia indicando ‘t’ en caso de que la necesite y ‘f’ en caso de que no la necesite
-  `	license: ` Este atributo indica el tipo de licencia necesaria en el caso de que lo sea, si no es necesaria, aparecerá vacía
-  `	jurisdiction_names: ` Nombre la jurisdicción aplicada al Airbnb, todos tienen la jurisdicción Whasington State Legislature
-  `	instant_bookable: ` Este atributo muestra ‘t’ en caso de que el Airbnb cuente con reserva instantánea y ‘f’ en caso contrario
-  `	cancellation_policy: ` Indica el tipo de política de cancelación
-  `	require_guest_profile_picture: ` Indica una ‘t’ si se requiere que el cliente tenga que tener una foto de perfil para realizar la reserva y ‘f’ en caso contrario
-  `	require_guest_phone_verification: ` Indica una ‘t’ si se requiere que el cliente tenga que tener verificado el número de teléfono y ‘f’ en caso contrario
-  `	calculated_host_listings_count: ` Este atributo muestra otra métrica para medir la experiencia del usuario (hay que revisarlo)
-  `	reviews_per_month: ` Número de valoraciones por semana, que tiene el Airbnb


In [3]:
names(airbnb_listings)

[1] "id"                               "listing_url"                     
 [3] "scrape_id"                        "last_scraped"                    
 [5] "name"                             "summary"                         
 [7] "space"                            "description"                     
 [9] "experiences_offered"              "neighborhood_overview"           
[11] "notes"                            "transit"                         
[13] "thumbnail_url"                    "medium_url"                      
[15] "picture_url"                      "xl_picture_url"                  
[17] "host_id"                          "host_url"                        
[19] "host_name"                        "host_since"                      
[21] "host_location"                    "host_about"                      
[23] "host_response_time"               "host_response_rate"              
[25] "host_acceptance_rate"             "host_is_superhost"               
[27] "host_thumbnail_url"               "host_picture_url"                
[29] "host_neighbourhood"               "host_listings_count"             
[31] "host_total_listings_count"        "host_verifications"              
[33] "host_has_profile_pic"             "host_identity_verified"          
[35] "street"                           "neighbourhood"                   
[37] "neighbourhood_cleansed"           "neighbourhood_group_cleansed"    
[39] "city"                             "state"                           
[41] "zipcode"                          "market"                          
[43] "smart_location"                   "country_code"                    
[45] "country"                          "latitude"                        
[47] "longitude"                        "is_location_exact"               
[49] "property_type"                    "room_type"                       
[51] "accommodates"                     "bathrooms"                       
[53] "bedrooms"                         "beds"                            
[55] "bed_type"                         "amenities"                       
[57] "square_feet"                      "price"                           
[59] "weekly_price"                     "monthly_price"                   
[61] "security_deposit"                 "cleaning_fee"                    
[63] "guests_included"                  "extra_people"                    
[65] "minimum_nights"                   "maximum_nights"                  
[67] "calendar_updated"                 "has_availability"                
[69] "availability_30"                  "availability_60"                 
[71] "availability_90"                  "availability_365"                
[73] "calendar_last_scraped"            "number_of_reviews"               
[75] "first_review"                     "last_review"                     
[77] "review_scores_rating"             "review_scores_accuracy"          
[79] "review_scores_cleanliness"        "review_scores_checkin"           
[81] "review_scores_communication"      "review_scores_location"          
[83] "review_scores_value"              "requires_license"                
[85] "license"                          "jurisdiction_names"              
[87] "instant_bookable"                 "cancellation_policy"             
[89] "require_guest_profile_picture"    "require_guest_phone_verification"
[91] "calculated_host_listings_count"   "reviews_per_month"

In [4]:
dim(airbnb_listings)

[1] 3818   92

Como podemos ver, el dataset tiene 92 atributos y 3818 filas. A continuación, vemos un ejemplo de una fila del dataset para obtener un primer vistazo sobre cómo se comportan los atributos.

In [5]:
airbnb_listings[1,]

id     listing_url                         scrape_id   last_scraped
1 241032 https://www.airbnb.com/rooms/241032 2.01601e+13 2016-01-04  
  name                         summary
1 Stylish Queen Anne Apartment        
  space                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
1 Make your self at home in this charming one-bedroom apartment, centrally-located on the west side of Queen Anne hill.   This elegantly-decorated, completely private apartment (bottom unit of a duplex) has an open floor plan, bamboo floors, a fully equipped kitchen, a TV,  DVD player, basic cable, and a very cozy bedroom with a queen-size bed. The unit sleeps up to four (two in the bedroom and two on the very comfortable fold out couch, linens included) and includes free WiFi and laundry. The apartment opens onto a private deck, complete with it's own BBQ, overlooking a garden and a forest of black bamboo.    The Apartment is perfectly-located just one block from the bus lines where you can catch a bus and be downtown Seattle in fifteen minutes or historic Ballard in ten or a quick five-minute walk will bring you to Whole Foods and Peet's Coffee or take a fifteen minute walk to the top of Queen Anne Hill where you will find a variety of eclectic shops, bars, and restaurants. There is no
  description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
1 Make your self at home in this charming one-bedroom apartment, centrally-located on the west side of Queen Anne hill.   This elegantly-decorated, completely private apartment (bottom unit of a duplex) has an open floor plan, bamboo floors, a fully equipped kitchen, a TV,  DVD player, basic cable, and a very cozy bedroom with a queen-size bed. The unit sleeps up to four (two in the bedroom and two on the very comfortable fold out couch, linens included) and includes free WiFi and laundry. The apartment opens onto a private deck, complete with it's own BBQ, overlooking a garden and a forest of black bamboo.    The Apartment is perfectly-located just one block from the bus lines where you can catch a bus and be downtown Seattle in fifteen minutes or historic Ballard

#### Calendario de reservas de Airbnb en Seattle

A continuación, haremos el mismo proceso con el conjunto de datos `airbnb_calendar`, el cual almacena información sobre todas las reservas realizadas en Airbnb sobre _hosts_ de Seattle.

In [6]:
names(airbnb_calendar)

[1] "listing_id" "date"       "available"  "price"

In [7]:
dim(airbnb_calendar)

[1] 1393570       4

Este dataset considera 1393570 filas con 4 atributos cada una de ellas con la siguiente interpretación:
- `listing_id`: id en el sistema de Airbnb del _listing_ al cual se corresponde la reserva.
- `date`: día, mes y año.
- `available`: disponibilidad del _listing_ en esta fecha.
- `price`: precio con el cual se oferta el _listing_ en esta fecha.

#### Reviews de Airbnb en Seattle

Realizamos el mismo proceso con el conjunto de datos `airbnb_reviews`, el cual contiene todas las reviews realizadas en Airbnb sobre _listings_ de Seattle.

In [8]:
names(airbnb_reviews)

[1] "listing_id"    "id"            "date"          "reviewer_id"  
[5] "reviewer_name" "comments"

In [9]:
dim(airbnb_reviews)

[1] 84849     6

El dataset considera X reviews realizadas; cada una de ellas contiene los siguientes atributos:
- `listing_id`: id en el sistema de Airbnb para el _listing_ sobre el cual se realiza la _review_.
- `id`: id en el sistema de Airbnb para la _review_.
- `date`: día, mes y año en el cual se envió la _review_ al sistema.
- `reviewer_id`: id en el sistema de Airbnb de la persona que escribió la _review_.
- `reviewer_name`: nombre de pila de la persona que escribió la _review_.
- `comments`: comentarios agregados a la _review_.

In [3]:
airbnb_reviews[1:5,]

listing_id id       date       reviewer_id reviewer_name
1 7202016    38917982 2015-07-19 28943674    Bianca       
2 7202016    39087409 2015-07-20 32440555    Frank        
3 7202016    39820030 2015-07-26 37722850    Ian          
4 7202016    40813543 2015-08-02 33671805    George       
5 7202016    41986501 2015-08-10 34959538    Ming         
  comments                                                                                                                                                                                                                                                                                                                                                                                              
1 Cute and cozy place. Perfect location to everything!                                                                                                                                                                                                                                                                                                                                                  
2 Kelly has a great room in a very central location. \nBeautiful building , architecture and a style that we really like. \nWe felt guite at home here and wish we had spent more time.\nWent for a walk and found Seattle Center with a major food festival in progress. What a treat.\nVisited the Space Needle and the Chihuly Glass exhibit. Then Pikes Place Market. WOW.  Thanks for a great stay.
3 Very spacious apartment, and in a great neighborhood.  This is the kind of apartment I wish I had!\n\nDidn't really get to meet Kelly until I was on my out, but she was always readily available by phone. \n\nI believe the only "issue" (if you want to call it that) was finding a place to park, but I sincerely doubt its easy to park anywhere in a residential area after 5 pm on a Friday    
4 Close to Seattle Center and all it has to offer - ballet, theater, museum, Space Needle, restaurants of all ilk just blocks away, and the Metropolitan (probably the coolest grocer you'll ever find). Easy to find and Kelly was warm, welcoming, and really interesting to talk to.                                                                                                                 
5 Kelly was a great host and very accommodating in a great neighborhood. She has some great coffee and while I wasn't around much during my stay the time I spent interacting with her was very pleasant. \n\nThe apartment is in a great location and very close to the Seattle Center. The neighborhood itself has a lot of good food as well!

#### Espacios culturales de Seattle

Procesamos con la misma metodología el conjunto de datos `cultural_spaces`, el cual nos muestra información sobre los espacios culturales de Seattle con los siguientes atributos:
- `	Name`: Nombre del Espacio Cultural
- `	Phone`: Número de teléono
- `	URL`: URL del espacio cultural
- `	Square Feet Total`: Metros cuadrados del espacio cultural
- `	Neighborhood`: Barrio en el que se encuentra ubicado el Espacio Cultural 
- `	Organization Type`: Tipo de Organización que organiza el evento
- `	Dominant Discipline`: Disciplina del Espacio cultural
- `	Year of Occupation`: Año durante el cual se encuentra disponible el Espacio Cultural en Seattle
- `	Rent vs Own`: Indica si el Espacio Cultural es alquilado o propio de la organización, teniendo de posibilidades Rent, Own u Other.
- `	Age of Current Building`: Años de antigüedad del Espacio Cultural
- `	Length of Lease (Date)`: Año hasta la cual dura el contrato de arrendamiento del Espacio Cultural, en el caso de que sea arrendado
- `	Year Organization Founded`: Año en el que fue fundada la Organización responsable del Espacio Cultural
- `	Number of Past Facilities`: Número de ediciones pasadas, en el caso de que las tenga
- `	Stages and Theaters`: Número de escenarios y teatros dentro del Espacio Cultural
- `	Stage & Theater Seats`: Número de butacas con las que cuenta el espacio cultural
- `	Gallery Space`: Galería con la que cuenta el Espacio Cultural
- `	Gallery Square Feet`: Metros cuadrados que tiene la galería del espacio cultural
- `	ADA Compilant`: Atributo que indica si el espacio cultural ofrece facilidades para personas con discapacidad
- `	Avaiable Parking`: Plazas de Parkng con las que cuenta el Espacio Cultural
- `	Street Presence`: Indica si el espacio cultural cuenta con presencia a la calle, lo que significa que el interior del mismo se puede ver desde la calle, captando así la atención del público
- `	Rental Space`: Indica si el espacio está arrendado o no 
- `	Alcohol Sales`: Indica si se vende alcohol en el Espacio Cultural
- `	Organization Mission`: Indica la misión de la organización del evento
- `	Founded by A&C`: Indica si el espacio fue fundado por A&C, una empresa de fabricación.
- `	Founded by 4Culture`: Indica si fue fundad por la autoridad 4Culture
- `	Stability Index (5=very stable, 1=very uncertain)`: Índice que indica la estabilidad del espacio cultural
- `	Control Index`: Índice que indica el control del espacio cultural
- `	Constituency over 50% one race`:
- `	Specific Demographics and Community`:
- `	Organization Leadership`:  Este atributo indica el lider de la organización del Espacio
- `	Closed datte`: Fecha en la que cierra el Espacio Cultural
- `	Closed`: Indica 1 si ha cerrado o 0 de lo contrario
- `	Address`: Dirección del Espacio Cultural
- `	Location`: Localización exacta del Espacio Cultural


In [4]:
names(cultural_spaces)

[1] "ï..Name"                                                      
 [2] "Phone"                                                        
 [3] "URL"                                                          
 [4] "Square.Feet.Total"                                            
 [5] "Neighborhood"                                                 
 [6] "Organization.Type"                                            
 [7] "Dominant.Discipline"                                          
 [8] "Year.of.Occupation"                                           
 [9] "Rent.vs.Own"                                                  
[10] "Age.of.Current.Building"                                      
[11] "Length.of.Lease..Date."                                       
[12] "Year.Organization.Founded"                                    
[13] "Number.of.Past.Facilities"                                    
[14] "Stages.and.Theaters"                                          
[15] "Stage...Theater.Seats"                                        
[16] "Gallery.Space"                                                
[17] "Gallery.Square.Feet"                                          
[18] "ADA.Compliant"                                                
[19] "Available.Parking"                                            
[20] "Street.Presence"                                              
[21] "Rental.Space"                                                 
[22] "Alcohol.Sales"                                                
[23] "Organization.Mission"                                         
[24] "Funded.by.A.C"                                                
[25] "Funded.by.4Culture"                                           
[26] "Stability.Index..5.very.stable..1.very.uncertain."            
[27] "Control.Index..5.very.in.control..1...very.out.of.control.."  
[28] "Constituency.over.50..one.race"                               
[29] "Specific.Demographics.and.Community"                          
[30] "Organization.Leadership"                                      
[31] "Organization.Artists"                                         
[32] "Closed.Date"                                                  
[33] "Closed."                                                      
[34] "Address"                                                      
[35] "Location"                                                     
[36] "X2010.Census.Tracts"                                          
[37] "Seattle.Police.Department.Micro.Community.Policing.Plan.Areas"
[38] "City.Council.Districts"                                       
[39] "SPD.Beats"                                                    
[40] "Zip.Codes"

In [5]:
dim(cultural_spaces)

[1] 1202   40

In [6]:
head(cultural_spaces, n=3)

ï..Name                  Phone          URL                    
1 Bootstrap Music                                                
2 Ghost Gallery            (206) 832-6063 www.ghostgalleryart.com
3 Chihuly Garden and Glass                                       
  Square.Feet.Total Neighborhood Organization.Type Dominant.Discipline
1  NA                                              Studios            
2 900               Capitol Hill N                 Visual             
3  NA                                              Visual             
  Year.of.Occupation Rent.vs.Own Age.of.Current.Building ...
1   NA                             NA                    ...
2 2010               R           1926                    ...
3   NA                             NA                    ...
  Organization.Artists Closed.Date Closed. Address                         
1                                  0                                       
2                                  0       504 E Denny Way Seattle WA 98122
3                                  0       305 Harrison St Seattle WA 98109
  Location                   X2010.Census.Tracts
1                            NA                 
2 (47.6186813, -122.3249251) NA                 
3 (47.6205686, -122.350491)  NA                 
  Seattle.Police.Department.Micro.Community.Policing.Plan.Areas
1 NA                                                           
2  8                                                           
3 50                                                           
  City.Council.Districts SPD.Beats Zip.Codes
1 NA                     NA           NA    
2 NA                      5        18376    
3 NA                     40        18390

### Exploración de los datos 

Con el objetivo de conocer cómo se comportan cada uno de los atributos de los dataset a utilizar y así facilitar y enriquecer las siguientes fases del proceso de Data Mining, realizaremos una exploración de los datos de cada uno de los datasets. Para ello, utilizaremos el paquete `psych`

In [7]:
if(!require(psych)){
    install.packages("psych")
    library(psych)
}

Loading required package: psych


#### Casas de Airbnb en Seattle

In [8]:
describe(airbnb_listings)

Warning message in FUN(newX[, i], ...):
"ningún argumento finito para min; retornando Inf"Warning message in FUN(newX[, i], ...):
"ningun argumento finito para max; retornando -Inf"

vars n    mean         sd          
id                                 1   3818 5.550111e+06 2.962660e+06
listing_url*                       2   3818 1.909500e+03 1.102306e+03
scrape_id                          3   3818 2.016010e+13 0.000000e+00
last_scraped*                      4   3818 1.000000e+00 0.000000e+00
name*                              5   3818 1.894389e+03 1.094537e+03
summary*                           6   3818 1.661389e+03 1.048870e+03
space*                             7   3818 1.318495e+03 1.000593e+03
description*                       8   3818 1.870324e+03 1.080956e+03
experiences_offered*               9   3818 1.000000e+00 0.000000e+00
neighborhood_overview*            10   3818 9.155990e+02 8.322909e+02
notes*                            11   3817 5.660375e+02 6.542211e+02
transit*                          12   3818 9.548080e+02 8.432268e+02
thumbnail_url*                    13   3818 1.603868e+03 1.081454e+03
medium_url*                       14   3818 1.603868e+03 1.081454e+03
picture_url*                      15   3818 1.909500e+03 1.102306e+03
xl_picture_url*                   16   3818 1.603868e+03 1.081454e+03
host_id                           17   3818 1.578556e+07 1.458382e+07
host_url*                         18   3818 1.407671e+03 8.077074e+02
host_name*                        19   3818 7.079995e+02 4.158170e+02
host_since*                       20   3818 7.187127e+02 3.918996e+02
host_location*                    21   3818 9.153405e+01 1.270384e+01
host_about*                       22   3818 8.079068e+02 6.724721e+02
host_response_time*               23   3818 4.981928e+00 1.115043e+00
host_response_rate*               24   3818 1.541697e+01 1.850057e+01
host_acceptance_rate*             25   3818 3.200629e+00 4.050760e-01
host_is_superhost*                26   3818 2.203248e+00 4.037679e-01
host_thumbnail_url*               27   3818 1.381509e+03 8.046569e+02
host_picture_url*                 28   3818 1.381509e+03 8.046569e+02
host_neighbourhood*               29   3818 4.302672e+01 3.265849e+01
host_listings_count               30   3816 7.157757e+00 2.862815e+01
...                               ...  ...  ...          ...         
guests_included                   63   3818   1.672603      1.3110396
extra_people*                     64   3818   7.264274      9.9432809
minimum_nights                    65   3818   2.369303     16.3059021
maximum_nights                    66   3818 780.447617   1683.5890075
calendar_updated*                 67   3818  21.882137      8.6575607
has_availability*                 68   3818   1.000000      0.0000000
availability_30                   69   3818  16.786276     12.1736368
availability_60                   70   3818  36.814825     23.3375414
availability_90                   71   3818  58.082504     34.0638454
availability_365                  72   3818 244.772656    126.7725263
calendar_last_scraped*            73   3818   1.000000      0.0000000
number_of_reviews                 74   3818  22.223415     37.7308923
first_review*                     75   3818 551.679937    335.9662422
last_review*                      76   3818 216.498952    112.1088870
review_scores_rating              77   3171  94.539262      6.6060832
review_scores_accuracy            78   3160   9.636392      0.6980312
review_scores_cleanliness         79   3165   9.556398      0.7972739
review_scores_checkin             80   3160   9.786709      0.5954990
review_scores_communication       81   3167   9.809599      0.5682114
review_scores_location            82   3163   9.608916      0.6290532
review_scores_value               83   3162   9.452245      0.7502586
requires_license*                 84   3818   1.000000      0.0000000
license                           85      0        NaN             NA
jurisdiction_names*               86   3818   1.000000      0.0000000
instant_bookable*                 87   3818   1.154793      0.3617547
cancellation_policy*              88   3818 

#### Calendario de reservas de Airbnb en Seattle

In [9]:
head(airbnb_calendar, n=5)

listing_id date       available price 
1 241032     2016-01-04 t         $85.00
2 241032     2016-01-05 t         $85.00
3 241032     2016-01-06 f               
4 241032     2016-01-07 f               
5 241032     2016-01-08 f

In [10]:
describe(airbnb_calendar)

vars n       mean         sd           median  trimmed     
listing_id 1    1393570 5.550111e+06 2.962274e+06 6118245 5.667756e+06
date*      2    1393570 1.830000e+02 1.053661e+02     183 1.830000e+02
available* 3    1393570 1.670610e+00 4.699919e-01       2 1.713263e+00
price*     4    1393570 2.072390e+02 2.537548e+02      61 1.767511e+02
           mad          min  max      range    skew       kurtosis 
listing_id 3391921.9320 3335 10340165 10336830 -0.3096599 -1.104326
date*          134.9166    1      365      364  0.0000000 -1.200021
available*       0.0000    1        2        1 -0.7260121 -1.472908
price*          88.9560    1      670      669  0.8323335 -1.032554
           se          
listing_id 2.509347e+03
date*      8.925578e-02
available* 3.981310e-04
price*     2.149561e-01

Como podemos ver, el atributo `available` se comporta como un booleano en el dataset. Además, el atributo `price` fluctúa entre 1$ y 670$ con mediana 61$.

In [11]:
nrow(airbnb_calendar[airbnb_calendar['available'] == 't',])

[1] 934542

Como podemos ver, en 2016 en Seattle hubo un total de X fechas en listings no reservadas.

In [12]:
nrow(airbnb_calendar[airbnb_calendar['listing_id'] == 241032 & airbnb_calendar['available'] == 't',])

[1] 346

La última ejecución nos muestra que 346 días del año 2016 el listing _241032_ no fue reservado, por lo tanto, tuvo una actividad muy baja. 

#### Reviews de Airbnb en Seattle

In [13]:
head(airbnb_reviews, n=5)

listing_id id       date       reviewer_id reviewer_name
1 7202016    38917982 2015-07-19 28943674    Bianca       
2 7202016    39087409 2015-07-20 32440555    Frank        
3 7202016    39820030 2015-07-26 37722850    Ian          
4 7202016    40813543 2015-08-02 33671805    George       
5 7202016    41986501 2015-08-10 34959538    Ming         
  comments                                                                                                                                                                                                                                                                                                                                                                                              
1 Cute and cozy place. Perfect location to everything!                                                                                                                                                                                                                                                                                                                                                  
2 Kelly has a great room in a very central location. \nBeautiful building , architecture and a style that we really like. \nWe felt guite at home here and wish we had spent more time.\nWent for a walk and found Seattle Center with a major food festival in progress. What a treat.\nVisited the Space Needle and the Chihuly Glass exhibit. Then Pikes Place Market. WOW.  Thanks for a great stay.
3 Very spacious apartment, and in a great neighborhood.  This is the kind of apartment I wish I had!\n\nDidn't really get to meet Kelly until I was on my out, but she was always readily available by phone. \n\nI believe the only "issue" (if you want to call it that) was finding a place to park, but I sincerely doubt its easy to park anywhere in a residential area after 5 pm on a Friday    
4 Close to Seattle Center and all it has to offer - ballet, theater, museum, Space Needle, restaurants of all ilk just blocks away, and the Metropolitan (probably the coolest grocer you'll ever find). Easy to find and Kelly was warm, welcoming, and really interesting to talk to.                                                                                                                 
5 Kelly was a great host and very accommodating in a great neighborhood. She has some great coffee and while I wasn't around much during my stay the time I spent interacting with her was very pleasant. \n\nThe apartment is in a great location and very close to the Seattle Center. The neighborhood itself has a lot of good food as well!

Exceptuando los ids y la fecha, todos los campos son textuales no estructurados (`reviewer_name` y `comments`), por lo tanto, la exploración se reduce en gran medida; no hay ninguna estructura que explorar.

#### Espacios culturales de Seattle

In [14]:
describe(cultural_spaces)

Warning message in FUN(newX[, i], ...):
"ningún argumento finito para min; retornando Inf"Warning message in FUN(newX[, i], ...):
"ningun argumento finito para max; retornando -Inf"

vars n   
ï..Name*                                                       1   1202
Phone*                                                         2   1202
URL*                                                           3   1202
Square.Feet.Total                                              4    666
Neighborhood*                                                  5   1202
Organization.Type*                                             6   1202
Dominant.Discipline*                                           7   1202
Year.of.Occupation                                             8    674
Rent.vs.Own*                                                   9   1202
Age.of.Current.Building                                       10    651
Length.of.Lease..Date.*                                       11   1202
Year.Organization.Founded                                     12    673
Number.of.Past.Facilities                                     13    656
Stages.and.Theaters                                           14    662
Stage...Theater.Seats                                         15    467
Gallery.Space*                                                16   1202
Gallery.Square.Feet                                           17    441
ADA.Compliant*                                                18   1202
Available.Parking                                             19    614
Street.Presence*                                              20   1202
Rental.Space*                                                 21   1202
Alcohol.Sales*                                                22   1202
Organization.Mission*                                         23   1202
Funded.by.A.C*                                                24   1202
Funded.by.4Culture*                                           25   1202
Stability.Index..5.very.stable..1.very.uncertain.             26    678
Control.Index..5.very.in.control..1...very.out.of.control..   27     29
Constituency.over.50..one.race                                28      0
Specific.Demographics.and.Community*                          29   1202
Organization.Leadership*                                      30   1202
Organization.Artists*                                         31   1202
Closed.Date*                                                  32   1202
Closed.                                                       33   1181
Address*                                                      34   1202
Location*                                                     35   1202
X2010.Census.Tracts                                           36     42
Seattle.Police.Department.Micro.Community.Policing.Plan.Areas 37    853
City.Council.Districts                                        38     37
SPD.Beats                                                     39    853
Zip.Codes                                                     40   1164
                                                              mean        
ï..Name*                                                      5.987072e+02
Phone*                                                        2.085216e+02
URL*                                                          2.100657e+02
Square.Feet.Total                                             3.003769e+04
Neighborhood*                                                 1.337105e+01
Organization.Type*                                            5.609817e+00
Dominant.Discipline*                                          1.243261e+01
Year.of.Occupation                                            1.979165e+03
Rent.vs.Own*                                                  2.608985e+00
Age.of.Current.Building                                       1.935845e+03
Length.of.Lease..Date.*                                       6.391015e+00
Year.Organization.Founded                                     1.984520e+03
Number.of.Past.Facilities                                     1.080793e+00
Stages.and.Theaters                                         

In [15]:
head(cultural_spaces, n=5)

ï..Name                  Phone          URL                    
1 Bootstrap Music                                                
2 Ghost Gallery            (206) 832-6063 www.ghostgalleryart.com
3 Chihuly Garden and Glass                                       
4 Dynamic Sound Service                                          
5 Golden Oldies                                                  
  Square.Feet.Total Neighborhood Organization.Type Dominant.Discipline
1  NA                                              Studios            
2 900               Capitol Hill N                 Visual             
3  NA                                              Visual             
4  NA                                              Studios            
5  NA                                              Music              
  Year.of.Occupation Rent.vs.Own Age.of.Current.Building ...
1   NA                             NA                    ...
2 2010               R           1926                    ...
3   NA                             NA                    ...
4   NA                             NA                    ...
5   NA                             NA                    ...
  Organization.Artists Closed.Date Closed. Address                         
1                                   0                                      
2                                   0      504 E Denny Way Seattle WA 98122
3                                   0      305 Harrison St Seattle WA 98109
4                                  NA                                      
5                                   0      201 NE 45th St Seattle WA 98105 
  Location                   X2010.Census.Tracts
1                            NA                 
2 (47.6186813, -122.3249251) NA                 
3 (47.6205686, -122.350491)  NA                 
4                            NA                 
5 (47.66116, -122.326927)    NA                 
  Seattle.Police.Department.Micro.Community.Policing.Plan.Areas
1 NA                                                           
2  8                                                           
3 50                                                           
4 NA                                                           
5 61                                                           
  City.Council.Districts SPD.Beats Zip.Codes
1 NA                     NA           NA    
2 NA                      5        18376    
3 NA                     40        18390    
4 NA                     NA           NA    
5 NA                      3        18383

In [16]:
tail(cultural_spaces, n=5)

ï..Name                                  Phone         
1198 Westminster Chapel                                     
1199 Woodinville Repertory Theatre            (206) 510-6304
1200 UMO Ensemble                             (206) 408-8059
1201 Washington Old Time Fiddlers Association               
1202 Valley Center Stage                      (425) 831-5667
     URL                          Square.Feet.Total Neighborhood
1198                              NA                            
1199 www.woodinvillerep.org       NA                            
1200 www.umo.org                  NA                            
1201                              NA                            
1202 www.intuitivelearningsys.org NA                            
     Organization.Type Dominant.Discipline Year.of.Occupation Rent.vs.Own
1198                   Preservation        NA                            
1199                   Performance         NA                            
1200                   Performance         NA                            
1201                   Music               NA                            
1202                   Performance         NA                            
     Age.of.Current.Building ... Organization.Artists Closed.Date Closed.
1198 NA                      ...                                  0      
1199 NA                      ...                                  0      
1200 NA                      ...                                  0      
1201 NA                      ...                                  0      
1202 NA                      ...                                  0      
     Address                                                 
1198 13646 NE 24th St Bellevue WA 98005                      
1199 16120 Redmond-Woodinville Rd NE #15 Woodinville WA 98072
1200 18870 SW 103rd Ave SW,  Vashon WA 98070                 
1201 18418 112th Avenue SE Renton WA 98055                   
1202 119 West North Bend Way North Bend WA 98045             
     Location                 X2010.Census.Tracts
1198 (47.63285, -122.157087)  NA                 
1199 (47.744744, -122.158388) NA                 
1200 (47.433761, -122.464157) NA                 
1201 (47.437241, -122.190124) NA                 
1202 (47.494937, -121.786457) NA                 
     Seattle.Police.Department.Micro.Community.Policing.Plan.Areas
1198 NA                                                           
1199 NA                                                           
1200 NA                                                           
1201 NA                                                           
1202 NA                                                           
     City.Council.Districts SPD.Beats Zip.Codes
1198 NA                     NA        31850    
1199 NA                     NA        18071    
1200 NA                     NA        18068    
1201 NA                     NA        17808    
1202 NA                     NA        17798

In [17]:
nrow(cultural_spaces)

[1] 1202

In [18]:
# Cuenta la cantidad de filas que no son NA
sum(is.na(cultural_spaces$Year.of.Occupation))

[1] 528

Lo primero que nos llama la atención es la gran cantidad de NAs que tiene el dataset en algunos campos que podrían ser interesantes, algo que se analizará con más precisión en el siguiente apartado correspondiente a calidad de los datos. La estructura de muchos atributos es confusa; a pesar de ello, encontramos que la localización (`Location`)está representada en un formato válido para nuestro análisis ya que expone la longitud y la latitud del espacio cultural, algo que nos será muy útil. Otro atributo que consideramos realmente útil es `Dominant.Discipline`, el cual tiene el siguiente comportamiento:

In [19]:
table(cultural_spaces$Dominant.Discipline)


                                         
                                       3 
Arts/Cultrual Administration or Advocacy 
                                       1 
Arts/Cultural Administration or Advocacy 
                                      39 
     Arts/Cultural Training or Education 
                                     143 
    Arts/Cultural Training or Education  
                                       1 
                                  Cinema 
                                      25 
                        Community Center 
                                      51 
                               Education 
                                      14 
                                Heritage 
                                      57 
                                Literary 
                                      78 
                               Live/Work 
                                      12 
                               Multi-use 
                                 

Encontramos múltiples categorías que podrían ser de utilidad en nuestro análisis; gracias a este atributo podemos conocer qué espacios culturales se dedican a la música, a la educación o son multiuso.

### Calidad de los datos 

La calidad de los datos de los datasets que tratamos la definiremos según la precisión, la completitud y la consistencia de sus datos. Para determinar qué calidad tienen los datos elegidos para la resolución de la cuestión de este proyecto, realizamos varios análisis en las siguientes líneas de código.

#### Casas de Airbnb en Seattle

#### Calendario de reservas de Airbnb en Seattle

#### Reviews de Airbnb en Seattle

In [20]:
nrow(airbnb_reviews)

[1] 84849

In [21]:
sum(is.na(airbnb_reviews$comments))

[1] 0

In [22]:
sum(airbnb_reviews$comments == "")

[1] 17

La calidad de los datos del dataset de reviews de casas de Airbnb de Seattle utilizado es muy buena, apenas 17 reviews de 84849 son inválidas.

#### Espacios culturales de Seattle

## Fase III. Data Preparation. Análisis de los datos y selección de características

Antes de comenzar el análisis de los datos de nuestros datasets, cargamos todos los paquetes necesarios para ello en las siguientes líneas de código. FALTA EXPLICACIÓN DE CADA PAQUETE.

In [30]:
if(!require(tidyverse)){
    install.packages("tidyverse")
    library(tidyverse)
}

if(!require(corrplot)){
    install.packages("corrplot")
   library(corrplot)
}

if(!require(dplyr)){
    install.packages("dplyr")
    library(dplyr)
}
if(!require(lattice)){
    install.packages("lattice")
    library(lattice) 
}
if(!require(ggplot2)){
    install.packages("ggplot2")
    library(ggplot2) 
}
if(!require(GGally)){
    install.packages("GGally")
    library(GGally)
}
if(!require(dplyr)){
    install.packages("dplyr")
    library(dplyr)
}

Loading required package: tidyverse
Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
"there is no package called 'tidyverse'"Installing package into 'C:/Users/franj/OneDrive/Documentos/R/win-library/3.5'
(as 'lib' is unspecified)
also installing the dependencies 'highr', 'markdown', 'xfun', 'rematch', 'knitr', 'tinytex', 'generics', 'DBI', 'ellipsis', 'cellranger', 'rmarkdown', 'selectr', 'broom', 'dbplyr', 'forcats', 'haven', 'lubridate', 'modelr', 'readr', 'readxl', 'reprex', 'rvest', 'tidyr', 'xml2'



package 'highr' successfully unpacked and MD5 sums checked
package 'markdown' successfully unpacked and MD5 sums checked
package 'xfun' successfully unpacked and MD5 sums checked
package 'rematch' successfully unpacked and MD5 sums checked
package 'knitr' successfully unpacked and MD5 sums checked
package 'tinytex' successfully unpacked and MD5 sums checked
package 'generics' successfully unpacked and MD5 sums checked
package 'DBI' successfully unpacked and MD5 sums checked
package 'ellipsis' successfully unpacked and MD5 sums checked
package 'cellranger' successfully unpacked and MD5 sums checked
package 'rmarkdown' successfully unpacked and MD5 sums checked
package 'selectr' successfully unpacked and MD5 sums checked
package 'broom' successfully unpacked and MD5 sums checked
package 'dbplyr' successfully unpacked and MD5 sums checked
package 'forcats' successfully unpacked and MD5 sums checked
package 'haven' successfully unpacked and MD5 sums checked
package 'lubridate' successfully

-- Attaching packages --------------------------------------- tidyverse 1.2.1 --
v ggplot2 3.1.1       v purrr   0.3.2  
v tibble  2.1.1       v dplyr   0.8.0.1
v tidyr   0.8.3       v stringr 1.4.0  
v readr   1.3.1       v forcats 0.4.0  
-- Conflicts ------------------------------------------ tidyverse_conflicts() --
x ggplot2::%+%()   masks psych::%+%()
x ggplot2::alpha() masks psych::alpha()
x dplyr::filter()  masks stats::filter()
x dplyr::lag()     masks stats::lag()
Loading required package: corrplot
corrplot 0.84 loaded
Loading required package: lattice
Loading required package: GGally

Attaching package: 'GGally'

The following object is masked from 'package:dplyr':

    nasa



In [24]:
version

               _                           
platform       x86_64-w64-mingw32          
arch           x86_64                      
os             mingw32                     
system         x86_64, mingw32             
status                                     
major          3                           
minor          5.3                         
year           2019                        
month          03                          
day            11                          
svn rev        76217                       
language       R                           
version.string R version 3.5.3 (2019-03-11)
nickname       Great Truth                 

### Seleccion de los datos y variables

Se analizará cada una de las columnas o variables de los diferentes dataset y su relevancia para nuestro problema para conocer qué variables es necesario eliminar ya que no aportan valor al posterior analisis y creación de modelos. Aunque en este apartado se realice una hipótesis sobre qué variables seleccionar, este conjunto de variables puede cambiar tras la construcción de modelos con el fin de obtener mejores resultados, si fuera necesario.

#### Casas de Airbnb en Seattle

En el caso del listado de casas de Airbnb, eliminamos todas aquellas variables que hemos considerado que no tienen relevancia en nuestro análisis, bien ya sea porque consideramos que su estado no afecta al precio o bien porque son variables que almacena Airbnb para el funcionamiento de su sistema, como por ejemplo la primera review que tuvo la casa, todos los ids que no fueran del listing, prácticamente todos los datos del perfil del host a excepción de si es superhost o no, ratios de actividad del host o urls a imágenes. Por lo tanto, se han mantenido diferentes variables que determinaban aspectos muy determinantes para el análisis propuesto, como todas aquellas que están relacionadas con las características explícitas de la casa. Además, se han tenido en cuenta variables con impacto indirecto sobre el problema, como la puntuación media las reviews por mes de la casa, ya que pueden ser muy interesantes. También eliminamos algunas variables que son redundantes, como aquellas de localización `country`, `state`, `city` o `country_code`, ya que en el contexto de nuestro problema no vamos a considerar otros países o ciudades; solo utilizaremos `zipcode` y `latitude`-`longitude` para conocer la localización de la casa, ya que podremos cruzar con el resto de datasets utilizando estos campos.

In [25]:
# Se seleccionan las variables que no se van tener en cuenta en la construcción del dataset
remove_listings <- c('listing_url','scrape_id','last_scraped','name','summary','space','description',
                     'experiences_offered','neighborhood_overview','notes','transit','thumbnail_url','medium_url',
                     'picture_url','xl_picture_url','host_id','host_url','host_name','host_since','host_location',
                     'host_about','host_response_time','host_response_rate','host_acceptance_rate',
                     'host_thumbnail_url','host_picture_url','host_neighbourhood', 'host_listings_count',
                     'host_total_listings_count','neighbourhood', 'neighbourhood_group_cleansed', 'is_location_exact',
                     'street','city','state', 'market','smart_location','country_code','country', 'security_deposit',
                     'guests_included', 'extra_people','minimum_nights','maximum_nights','has_availability',
                     'calendar_updated','availability_30','availability_60','availability_90','availability_365',
                     'review_scores_accuracy', 'requires_license', 'require_guest_phone_verification',
                     'calculated_host_listings_count','calendar_last_scraped','first_review',
                     'last_review','license','jurisdiction_names','host_verifications', 'weekly_price','monthly_price',
                     'instant_bookable','cancellation_policy','require_guest_profile_picture')

airbnb_listings_pre <- airbnb_listings[ , !(names(airbnb_listings) %in% remove_listings)]

El dataset ha quedado de la siguiente manera:

In [26]:
names(airbnb_listings_pre)

[1] "id"                          "host_is_superhost"          
 [3] "host_has_profile_pic"        "host_identity_verified"     
 [5] "neighbourhood_cleansed"      "zipcode"                    
 [7] "latitude"                    "longitude"                  
 [9] "property_type"               "room_type"                  
[11] "accommodates"                "bathrooms"                  
[13] "bedrooms"                    "beds"                       
[15] "bed_type"                    "amenities"                  
[17] "square_feet"                 "price"                      
[19] "cleaning_fee"                "number_of_reviews"          
[21] "review_scores_rating"        "review_scores_cleanliness"  
[23] "review_scores_checkin"       "review_scores_communication"
[25] "review_scores_location"      "review_scores_value"        
[27] "reviews_per_month"

#### Calendario de reservas de Airbnb en Seattle

En el caso del calendario se ha procedido a mantener todo el conjunto de variables, dado que la cantidad de variables es muy pequeña y todas ellas pueden tener un impacto positivo sobre el análisis.

In [ ]:
airbnb_calendar_pre <- airbnb_calendar

El dataset ha quedado de la siguiente manera:

In [ ]:
names(airbnb_calendar_pre)

#### Reviews de Airbnb en Seattle

La utilidad de este dataset en nuestro análisis se ve limitada ya que no realizaremos procesamiento de textos sobre el atributo `comments`, único atributo de valor del dataset, debido a que escapa de nuestros objetivos. Además, tampoco utilizaremos este dataset para realizar agregaciones como haremos con el resto de datasets, ya que el dataset _listings_ ya contiene todoa la información sobre reviews que nos gustaría tratar. Por ello, decidimos no continuar utilizando este dataset de ahora en adelante en nuestra implementación. La siguiente línea borra el dataset cargado para aligerar la memoria y facilitar los próximos cálculos.

In [ ]:
rm(airbnb_reviews)

#### Espacios culturales de Seattle

En el caso de los espacios culturales consideramos importante seleccionar las variables que son derivadas de la localizacion, determinante para cruzarlas con la localizacion del airbnb que queremos estudiar. Para ello, consideraremos la localización por `Location`, ya que almacena la latitud y longitud que posteriormente extraeremos. Trabajaremos con latitud y longitud en este dataset tanto el dataset _listings_ como este dataset contiene esta información.  Además, seleccionamos también el atributo `Dominant.Discipline` ya que en los siguientes apartados realizaremos agregaciones con él debido a su gran valor. En las siguientes líneas de código borramos todos los campos que no sean estos. Consideramos únicamente estos campos por los motivos mencionados además de la baja calidad del resto de campos.

In [63]:
remove_spaces <- c('Name','Phone','URL','Square.Feet.Total','Neighborhood','Organization.Type',
                   'Year.of.Occupation','Rent.vs.Own','Age.of.Current.Building','Length.of.Lease..Date.',
                   'Year.Organization.Founded','Number.of.Past.Facilities','Stages.and.Theaters',
                   'Stage...Theater.Seats','Gallery.Space','Gallery.Square.Feet','ADA.Compliant',
                   'Available.Parking','Street.Presence','Rental.Space','Alcohol.Sales',
                   'Organization.Mission','Funded.by.A.C','Funded.by.4Culture',
                   'Stability.Index..5.very.stable..1.very.uncertain.',
                   'Control.Index..5.very.in.control..1...very.out.of.control..',
                   'Constituency.over.50..one.race','Specific.Demographics.and.Community',
                   'Organization.Leadership','Organization.Artists','Closed.Date','Closed.',
                   'Address','Zip.Codes','X2010.Census.Tracts',
                   'Seattle.Police.Department.Micro.Community.Policing.Plan.Areas',
                   'City.Council.Districts','SPD.Beats')

cultural_spaces_pre <- cultural_spaces[ , !(names(cultural_spaces) %in% remove_spaces)]

El dataset ha quedado de la siguiente manera:

In [64]:
names(cultural_spaces_pre) 

[1] "Dominant.Discipline" "Zip.Codes"           "Latitude"           
[4] "Longitude"

### Limpieza de datos

In [65]:
head(airbnb_listings_pre, n=5)

ERROR: Error in head(airbnb_listings_pre, n = 5): objeto 'airbnb_listings_pre' no encontrado


In [ ]:
head(airbnb_calendar_pre, n=5)

In [ ]:
df <- select(airbnb_listings, host_response_rate, host_acceptance_rate, property_type,room_type,accommodates,bathrooms,bedrooms,
               beds,price,number_of_reviews,review_scores_rating,cancellation_policy, reviews_per_month)

In [ ]:
head(df)

In [ ]:
# Numbers of NaN values
sapply(df, function(df) sum(is.na(df)))

In [ ]:
# Number of rows before omit NaN
count(df)

In [ ]:
df2 <- na.omit(df)

In [ ]:
# Number of rows after omit NaN
nrow(df2)

In [ ]:
levels(df2$property_type)

In [ ]:
levels(df2$room_type)

In [ ]:
levels(df2$cancellation_policy)

In [ ]:
df2$property_type = factor(df2$property_type, levels = levels(df2$property_type), labels = seq(1,17))

In [ ]:
df2$room_type = factor(df2$room_type, levels = levels(df2$room_type), labels = seq(1,3))

In [ ]:
df2$cancellation_policy = factor(df2$cancellation_policy, levels = levels(df2$cancellation_policy), labels = seq(1,3))

In [ ]:
head(df2)

In [ ]:
levels(df2$host_response_rate)[length(levels(df2$host_response_rate))] <- "0%"
levels(df2$host_acceptance_rate)[length(levels(df2$host_acceptance_rate))] <- "0%"

In [ ]:
levels(df2$host_response_rate)

In [ ]:
levels(df2$host_response_rate) <- gsub(".$", "", levels(df2$host_response_rate))
levels(df2$host_acceptance_rate) <- gsub(".$", "", levels(df2$host_acceptance_rate))
levels(df2$price) <- substring(levels(df2$price), 2)

In [ ]:
for(i in c(1,3:ncol(df2))) {
    df2[,i] <- as.numeric(df2[,i])
}
df2$host_acceptance_rate <- as.numeric(df2$host_acceptance_rate)

In [ ]:
lapply(df2, class)

In [ ]:
M = cor(df2)
corrplot(M, method = "circle") #plot matrix

In [ ]:
hist(df2$price)

In [ ]:
plot(df2$price, df2$bedrooms)

In [ ]:
plot(df2$number_of_reviews, df2$price)

In [ ]:
pairs(df2)

In [ ]:
pairs(df2[1:7])

In [ ]:
pairs(df2[8:15])

In [ ]:
ggpairs(df2) 

### Estruturado de datos

### Integración de los datos

El objetivo de este apartado es integrar todos los datasets, cada uno con su interpretabilidad deseada en un único dataset. Tomaremos de base el dataset de casas de Airbnb sobre el cual agregaremos los datos del resto de datasets; para ello, realizamos los siguientes pasos.

#### Integración con  calendario de Airbnb

Añadimos al dataset de listings, la variable que nos indica una media de reservas al mes, para cuyo cálculo hay que recurrir al dataset calendar.

In [ ]:
booking_per_month=vector(mode='numeric',length=3818)

In [ ]:
airbnb_listings_def = data.frame(airbnb_listings_pre,booking_per_month)

In [ ]:
max = nrow(airbnb_listings_def)

In [ ]:
for(i in 1:max){
    id = airbnb_listings_def[c(i:i),c(1:1)]
    airbnb_listings_def[c(i:i),c(53:53)] = round((nrow(airbnb_calendar[airbnb_calendar['listing_id'] == id & airbnb_calendar['available'] == 'f',]))/12,digits=0)
}

In [ ]:
head(airbnb_listings_def, n=5)

#### Integración con  espacios culturales de Seattle

Introduccimos lo necesario para integrar el calculador de distancias entre dos puntos, en este caso entre un host de airbnb y un espacio cultural.

In [94]:
head(cultural_spaces$Location)

[1]                            (47.6186813, -122.3249251)
[3] (47.6205686, -122.350491)                            
[5] (47.66116, -122.326927)    (47.6142449, -122.1934783)
1101 Levels:  (47.188389, -122.008777) ... (47.805356, -122.148902)

In [77]:
cultural_spaces['Latitude'] <- NA #3
cultural_spaces['Longitude'] <- NA #4

for (i in 1:nrow(cultural_spaces)){
    loc = unlist(strsplit(as.character(cultural_spaces[i,35]), ","))
    if(length(loc)>0){
        loc_parse <- gsub("[^0-9.-]", "", loc)
        cultural_spaces[i,41] <- as.numeric(loc_parse[1])
        cultural_spaces[i,42] <- as.numeric(loc_parse[2])
    }
}

In [81]:
cultural_spaces$Latitude[2]

[1] 47.61868

In [89]:
if(!require(geosphere)){
    install.packages("geosphere")
    library(geosphere)
}

In [91]:
# Test
distm(c(cultural_spaces$Longitude[3], cultural_spaces$Latitude[3]), c(airbnb_listings_pre$longitude[1], airbnb_listings_pre$latitude[1]), fun = distHaversine)

[,1]    
[1,] 2331.453

Distancia máxima en metros en la cual se tendran en cuenta los espacios culturales y se incrementaran. 

In [92]:
max_distance <- 3000

In [104]:
airbnb_listings_pre['Num.Cultural.Spaces'] <- NA #3

In [123]:
head(airbnb_listings_pre[7])

latitude
1 47.63629
2 47.63912
3 47.62972
4 47.63847
5 47.63292
6 47.63053

In [28]:
#for (i in 1:nrow(airbnb_listings_pre)){
#    for (j in 1:nrow(cultural_spaces)){
#        print(cultural_spaces[j,42])
#        distm(c(cultural_spaces[j,42], cultural_spaces[j,43]), c(airbnb_listings_pre[i,8], airbnb_listings_pre[i,7]), fun = distHaversine)
#    }
#    
#}

In [29]:
must_convert <- sapply (airbnb_listings_pre,is.factor)

In [30]:
M2<-sapply(airbnb_listings_pre[,must_convert],unclass)

In [31]:
out<-cbind(airbnb_listings_pre[,!must_convert],M2)

### Formateo de los datos

A continuación, realizaremos el formateo de los datos del dataset obtenido en el paso anterior con el objetivo de dar como salida de esta fase un dataset apropiado para la construcción de un modelo de predicción sobre él.

## Fase IV. Modeling. Modelado

Una vez hemos preparado el dataset que utilizaremos en la construcción de nuestros modelo predictivo, iniciamos el modelado requerido para obtener dicho modelo.

### Técnica de modelado

FALTA EXPLICACIÓN DE ESTE APARTADO

ENLACES
Prediciendo precios de casas con regresion:
https://towardsdatascience.com/machine-learning-project-predicting-boston-house-prices-with-regression-b4e47493633d


https://www.datacamp.com/community/tutorials/linear-regression-R

https://www.r-bloggers.com/15-types-of-regression-you-should-know/

### Plan de prueba

A continuación, dividimos el dataset `df` en dos: datos de entrenamiento y datos de testing. De esta forma, evaluamos nuestro modelo predictivo sobre datos diferentes a los que ha utilizado para entrenar, es decir, datos nuevos. Para ello, dividiremos el dataset en 80% para datos de entrenamiento y el 20% en datos de test.

In [4]:
if(!require(caTools)){
    install.packages("caTools")
    library(caTools)
}

Loading required package: caTools
Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘caTools’”

In [ ]:
set.seed(123) # Con la misma semilla siempre tendremos los mismos datos generados
split_sample = sample.split(df,SplitRatio = 0.8) 
df_train =subset(df, split_sample ==TRUE)
df_test =subset(df, split_sample==FALSE)

### Construcción del modelo

Construimos un modelo de regresión lineal utilizando como variables características todas las variables del dataframe a excepción de la variable `price`, la cual será la variable objetivo.

In [ ]:
lm_fit = lm(price ~ ., data=df_train)

### Evaluación del modelo

In [ ]:
summary(lm_fit)

In [ ]:
plot(lm_fit$residuals, pch = 16, col = "red")

Para conocer el rendimiento del modelo utilizaremos la métrica _$ R^2$ score_, la cual razona un buen rendimiento con un valor cercano a 1 y un mal rendimiento con un valor cercano a 0. En este caso, nos enfocaremos en el valor _$ R^2$ score ajustado_, el cual funciona mejor con modelos con múltiples variables características, como el nuestro.

La mejor regresion lineal posible:
https://stackoverflow.com/questions/33740012/command-for-finding-the-best-linear-model-in-r

In [ ]:
# Stepwise Regression
if(!require(MASS)){
    install.packages("MASS")
    library(MASS)
}
fit <- lm(y~x1+x2+x3,data=mydata)
step <- stepAIC(fit, direction="both")
step$anova # display results 

## Fase V. Evaluation. Evaluación. Obtención de resultados.